### PR TITLE
Suppress `ref readonly` warning for extension method receiver

### PR DESCRIPTION
**Pull request by jjonescz**
Speclet: [the relevant section (second paragraph)](https://github.com/dotnet/csharplang/blob/3daa6f86510906df8ad29875659d0359d1a88b94/proposals/ref-readonly-parameters.md#overload-resolution) · [PR](https://github.com/dotnet/csharplang/pull/7350)
Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -3300,11 +3301,14 @@ private BoundExpression BindArgumentExpression(BindingDiagnosticBag diagnostics,
                             }
                             else if (this.CheckValueKind(argument.Syntax, argument, BindValueKind.Assignable, checkingReceiver: false, BindingDiagnosticBag.Discarded))
                             {
-                                // Argument {0} should be passed with 'ref' or 'in' keyword
-                                diagnostics.Add(
-                                    ErrorCode.WRN_ArgExpectedRefOrIn,
-                                    argument.Syntax,
-                                    arg + 1);
+                                if (!invokedAsExtensionMethod)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3304]: >if (!invokedAsExtensionMethod) [](http://example.com/codeflow?start=32&length=30)

Suppressing for all arguments? Looks suspicious.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 2]
```diff
@@ -3045,6 +3045,68 @@ .maxstack  2
             """);
     }
 
+    [Theory, CombinatorialData]
+    public void Invocation_ExtensionMethod([CombinatorialValues("ref readonly", "ref", "in")] string modifier)
+    {
+        var source = $$"""
+            static class C
+            {
+                static void M(this {{modifier}} int x) => System.Console.Write(x);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:3054]: >this {{modifier}} int x) [](http://example.com/codeflow?start=30&length=24)

Are we testing mismatch on other parameters?

**Comment by AlekseyTs** [comment]
>                                     ErrorCode.WRN_RefReadonlyNotVariable,

Are we testing this code path for this parameter of an extension method?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3298 in a941da0. [](commit_id = a941da003cff6492af36ee698035c4f5920bd32c, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                     ErrorCode.WRN_ArgExpectedIn,

Are we testing this code path for this parameter of an extension method?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3317 in a941da0. [](commit_id = a941da003cff6492af36ee698035c4f5920bd32c, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by AlekseyTs** [comment]
>                                     arg + 1);

Here and below we probably shouldn't do ```+ 1``` when ```invokedAsExtensionMethod``` is true. Otherwise we are getting confusing messages like:
```
            // (9,14): warning CS9503: Argument 2 should be passed with 'ref' or 'in' keyword
            //         x.M1(x);
```

But there is no second argument.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3288 in 596600a. [](commit_id = 596600a66e7f187b225ab05c6c53297f5ead3edb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

**Comment by jjonescz** [comment]
@jcouv for a second review, thanks

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Pull request merged**